File: extensions/smart-bundle-transform/src/selection.js
import { MAX_COMPONENT_QUANTITY, MAX_SELECTION_BYTES } from "./constants.js";
import { hasExactKeys, isObject, isSelectorKey, isVariantGid } from "./validation.js";

export function readSelection(value, runtime) {
  const input = parseAttribute(value);
  if (!input || input.rv !== runtime.rv) return null;
  const pairs = readPairs(input.s, runtime.selectors.length);
  if (!pairs) return null;
  const components = resolveComponents(pairs, runtime);
  return components ? aggregateComponents(components) : null;
}

function parseAttribute(value) {
  if (typeof value !== "string" || value.length > MAX_SELECTION_BYTES) return null;
  try {
    const parsed = JSON.parse(value);
    return isObject(parsed) && hasExactKeys(parsed, ["rv", "s"]) ? parsed : null;
  } catch {
    return null;
  }
}

function readPairs(value, expectedLength) {
  if (!Array.isArray(value) || value.length !== expectedLength) return null;
  const pairs = value.map(readPair);
  if (!pairs.every(Boolean)) return null;
  return new Set(pairs.map(({ key }) => key)).size === pairs.length ? pairs : null;
}

function readPair(value) {
  if (!Array.isArray(value) || value.length !== 2) return null;
  if (!isSelectorKey(value[0]) || !isVariantGid(value[1])) return null;
  return { key: value[0], variantId: value[1] };
}

function resolveComponents(pairs, runtime) {
  const byKey = new Map(pairs.map((pair) => [pair.key, pair.variantId]));
  const resolved = runtime.selectors.map((selector) => resolveSelector(selector, byKey, runtime.components));
  return resolved.every(Boolean) ? resolved : null;
}

function resolveSelector(selector, byKey, components) {
  const selectedId = byKey.get(selector.key);
  if (!selectedId) return null;
  const matches = selector.options.filter((index) => components[index].variantId === selectedId);
  if (matches.length !== 1) return null;
  return { ...components[matches[0]], discountPercent: selector.discountPercent };
}

function aggregateComponents(components) {
  const totals = new Map();
  for (const component of components) {
    const key = `${component.variantId}\u0000${component.discountPercent}`;
    const current = totals.get(key);
    if (current && current.unitPrice !== component.unitPrice) return null;
    const quantity = (current?.quantity || 0) + component.quantity;
    if (quantity > MAX_COMPONENT_QUANTITY) return null;
    totals.set(key, { ...component, quantity });
  }
  return [...totals.values()].map(({ variantId, ...component }) => ({ merchandiseId: variantId, ...component }));
}

File: extensions/smart-bundle-transform/src/selection.rs
use crate::{money, validation::*, BundleResult, Json};
use serde_json::json;
use std::collections::{BTreeMap, BTreeSet};

pub fn resolve(line: &crate::input::line::Line, runtime: &Json) -> BundleResult<Vec<Json>> {
    let selection = attribute(line)?;
    if !exact_keys(&selection, &["s"]) { return Err("Invalid bundle selection."); }
    let selectors = array(&runtime["s"])?;
    selector_count(selectors.len())?;
    let pairs = selected_pairs(&selection["s"], selectors.len())?;
    let mut totals = BTreeMap::new();
    let mut keys = BTreeSet::new();
    for selector in selectors {
        let key = integer(&selector["k"])?;
        if !keys.insert(key) { return Err("Duplicate bundle selector."); }
        let selected = pairs.get(&key).ok_or("Missing bundle component.")?;
        let component = resolve_option(selector, selected, array(&runtime["c"])?)?;
        let total = totals.entry(selected.to_string()).or_insert(0u64);
        *total += quantity(&component[1])?;
        if *total > 2000 { return Err("Bundle quantity exceeds Shopify's limit."); }
    }
    Ok(totals.into_iter().map(|(id, quantity)| json!({ "merchandiseId": id, "quantity": quantity })).collect())
}

fn selected_pairs(value: &Json, count: usize) -> BundleResult<BTreeMap<u64, &str>> {
    let pairs = array(value)?;
    if pairs.len() != count { return Err("Missing bundle selection."); }
    let mut selected = BTreeMap::new();
    for pair in pairs {
        if array(pair)?.len() != 2 { return Err("Invalid bundle selection."); }
        if selected.insert(integer(&pair[0])?, variant(&pair[1])?).is_some() { return Err("Duplicate selection."); }
    }
    Ok(selected)
}

fn resolve_option<'a>(selector: &Json, selected: &str, components: &'a [Json]) -> BundleResult<&'a Json> {
    if money::percent(&selector["d"])? != 0 { return Err("Invalid fixed bundle discount."); }
    let options = array(&selector["o"])?;
    if options.is_empty() { return Err("Empty bundle selector."); }
    let mut ids = BTreeSet::new();
    let mut resolved = None;
    for index in options {
        let component = components.get(integer(index)? as usize).ok_or("Invalid bundle option.")?;
        let id = format!("gid://shopify/ProductVariant/{}", text(&component[0])?);
        if !ids.insert(id.clone()) { return Err("Duplicate bundle variant."); }
        if id == selected { resolved = Some(component); }
    }
    resolved.ok_or("Variant is not allowed in this bundle.")
}
